src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs

```diff
@@ -0,0 +1,506 @@
+            }
+        }
+
+        private SafeProvHandle SafeProvHandle
+        {
+            get
+            {
+                if (_safeProvHandle == null)
+                {
+                    lock (_parameters)
+                    {
+                        if (_safeProvHandle == null)
+                        {
+                            SafeProvHandle hProv = CapiHelper.CreateProvHandle(_parameters, _randomKeyContainer);
+
+                            Debug.Assert(hProv != null);
+                            Debug.Assert(!hProv.IsInvalid);
+                            Debug.Assert(!hProv.IsClosed);
+
+                            _safeProvHandle = hProv;
+                        }
+                    }
+
+                    return _safeProvHandle;
+                }
+
+                return _safeProvHandle;
+            }
+            set
+            {
+                lock (_parameters)
+                {
+                    SafeProvHandle current = _safeProvHandle;
+
+                    if (ReferenceEquals(value, current))
+                    {
+                        return;
+                    }
+
+                    if (current != null)
+                    {
+                        SafeKeyHandle keyHandle = _safeKeyHandle;
+                        _safeKeyHandle = null;
+                        keyHandle?.Dispose();
+                        current.Dispose();
+                    }
+
+                    _safeProvHandle = value;
+                }
+            }
+        }
+
+        private SafeKeyHandle SafeKeyHandle
+        {
+            get
+            {
+                if (_safeKeyHandle == null)
+                {
+                    lock (_parameters)
+                    {
+                        if (_safeKeyHandle == null)
+                        {
+                            SafeKeyHandle hKey = CapiHelper.GetKeyPairHelper(
+                                CapiHelper.CspAlgorithmType.Dss,
+                                _parameters,
+                                _keySize,
+                                SafeProvHandle);
+
+                            Debug.Assert(hKey != null);
+                            Debug.Assert(!hKey.IsInvalid);
+                            Debug.Assert(!hKey.IsClosed);
+
+                            _safeKeyHandle = hKey;
+                        }
+                    }
+                }
+
+                return _safeKeyHandle;
+            }
+
+            set
+            {
+                lock (_parameters)
+                {
+                    SafeKeyHandle current = _safeKeyHandle;
+
+                    if (ReferenceEquals(value, current))
+                    {
+                        return;
+                    }
+
+                    _safeKeyHandle = value;
+                    current?.Dispose();
+                }
+            }
+        }
+
+        /// <summary>
+        /// CspKeyContainerInfo property
+        /// </summary>
```

Comment:
I agree. Those documents were copy-paste from RSA side... I'll change to the msdn info